query-engine/core/src/query_graph_builder/extractors/filters/mod.rs

```diff
@@ -38,28 +38,61 @@ pub fn extract_unique_filter(value_map: ParsedInputMap, model: &ModelRef) -> Que
                             field_name, model.name
                         ))
                     })
-                    .and_then(|fields| handle_compound_field(fields, value)),
+                    .and_then(|fields| handle_compound_field(model, fields, value)),
             }
         })
         .collect::<QueryGraphBuilderResult<Vec<Filter>>>()?;
 
     Ok(Filter::and(filters))
 }
 
-fn handle_compound_field(fields: Vec<ScalarFieldRef>, value: ParsedInputValue) -> QueryGraphBuilderResult<Filter> {
+fn handle_compound_field(
+    model: &ModelRef,
+    path_fields: Vec<(Vec<String>, ScalarFieldRef)>,
+    value: ParsedInputValue,
+) -> QueryGraphBuilderResult<Filter> {
     let mut input_map: ParsedInputMap = value.try_into()?;
-
-    let filters: Vec<Filter> = fields
-        .into_iter()
-        .map(|sf| {
-            let pv: PrismaValue = input_map.remove(&sf.name).unwrap().try_into()?;
-            Ok(sf.equals(pv))
-        })
-        .collect::<QueryGraphBuilderResult<Vec<_>>>()?;
+    let mut filters = Vec::with_capacity(path_fields.len());
+
+    for (path, field) in path_fields {
+        if path.len() > 1 {
```

Comment:
Do we need this `path.len() > 1` check here? We do the same check again in the `tracer_composite_filter` and the same code is repeated. We could probably let the `tracer_composite_filter` do the check

Reply:
Yes you do, this is an indirect recursive function and without it you don't have an exit case.